### PR TITLE
ci: unflake "go perf"

### LINKED ISSUE #3272: ci: "Go Perf" contains flaky test
When contributing a PR, you may see this:

image

and the logs contain nothing because they're too spammy (that's a different issue). Downloading the logs you find this:

```
2021-03-15T10:57:23.2270260Z goos: linux
2021-03-15T10:57:23.2270861Z goarch: amd64
2021-03-15T10:57:23.2272162Z pkg: github.com/open-policy-agent/opa/test/e2e/wasm/authz
2021-03-15T10:57:24.6422076Z BenchmarkRESTAuthzForbidAuthn-2            --- FAIL: BenchmarkRESTAuthzForbidAuthn-2
2021-03-15T10:57:24.6423674Z     authz_bench_integration_test.go:156: unexpected response status: 500 Internal Server Error
```

This is due to #3211. At least according to my current theory, we're having a race there between the HTTP client's request's context being closed and the wasm VM having finished its processing.

The plan in #3211 was to check for spurious interrupts from previous invocations, and not fail because of them, but either some code path was missed, or there's more to do to cleanup.

### PR DESCRIPTION
**Pull request by srenatus**
Now, when an interrupt happens, we'll clean up after ourselves: we keep calling
a cheap function to ensure that the trap has been trapped on.

To get there, we'll move the "defer-recover" further down the call stack.

👉 Also, this changes the cancellation error returned by the topdown builtin. It no
longer is a builtinError with a message indicating that the context was cancelled
(or its deadline reached), but return a CancelErr, the same thing that happens via
the other cancellation mechanisms involved topdown's Cancel (like in cidr.expand).

Compared with master, this isn't worse:

    name                       old time/op    new time/op    delta
    RESTAuthzForbidAuthn-16       542µs ±15%     525µs ±12%    ~     (p=0.841 n=5+5)
    RESTAuthzForbidPath-16        818µs ± 2%     827µs ± 2%    ~     (p=0.556 n=4+5)
    RESTAuthzForbidMethod-16      864µs ± 2%     851µs ± 4%    ~     (p=0.310 n=5+5)
    RESTAuthzAllow10Paths-16      896µs ±20%     855µs ± 5%    ~     (p=1.000 n=5+5)
    RESTAuthzAllow100Paths-16    4.28ms ± 3%    4.07ms ± 3%  -4.97%  (p=0.008 n=5+5)

    name                       old alloc/op   new alloc/op   delta
    RESTAuthzForbidAuthn-16      68.8kB ± 1%    67.2kB ± 1%  -2.28%  (p=0.008 n=5+5)
    RESTAuthzForbidPath-16       68.5kB ± 0%    66.9kB ± 0%  -2.33%  (p=0.008 n=5+5)
    RESTAuthzForbidMethod-16     68.5kB ± 0%    66.9kB ± 0%  -2.33%  (p=0.008 n=5+5)
    RESTAuthzAllow10Paths-16     68.5kB ± 0%    66.9kB ± 0%  -2.33%  (p=0.008 n=5+5)
    RESTAuthzAllow100Paths-16    69.1kB ± 0%    67.5kB ± 0%  -2.31%  (p=0.008 n=5+5)

    name                       old allocs/op  new allocs/op  delta
    RESTAuthzForbidAuthn-16       1.73k ± 1%     1.64k ± 1%  -5.17%  (p=0.008 n=5+5)
    RESTAuthzForbidPath-16        1.72k ± 0%     1.63k ± 0%    ~     (p=0.079 n=4+5)
    RESTAuthzForbidMethod-16      1.72k ± 0%     1.63k ± 0%  -5.13%  (p=0.008 n=5+5)
    RESTAuthzAllow10Paths-16      1.72k ± 0%     1.63k ± 0%  -5.13%  (p=0.008 n=5+5)
    RESTAuthzAllow100Paths-16     1.72k ± 0%     1.63k ± 0%  -5.16%  (p=0.008 n=5+5)

👉 Also, we're threading through some context arguments we hadn't had before. (See the two commits.)

👉 Also, I've added what I believe to be a missing return stmt in the handler code, please see the comment on the code added below.

------
Reproducing the flaky test behaviour locally using docker desktop goes as follows,

1. throttle your docker performance (I've used 2 CPU, 4G memory)
2. edit your makefile:
```diff
diff --git a/Makefile b/Makefile
index d16dc2f0..de26c42e 100644
--- a/Makefile
+++ b/Makefile
@@ -109,6 +109,10 @@ test-coverage: generate
 perf: generate
        $(GO) test $(GO_TAGS),slow $(GO_TEST_TIMEOUT) -run=- -bench=. -benchmem ./...

+.PHONY: flake
+flake:
+       while $(GO) test $(GO_TAGS),slow $(GO_TEST_TIMEOUT) -run=- -bench=. -benchmem ./test/e2e/wasm/authz ; do echo ; done;
+
 .PHONY: perf-noisy
 perf-noisy: generate
        $(GO) test $(GO_TAGS),slow,noisy $(GO_TEST_TIMEOUT) -run=- -bench=. -benchmem ./...
```
3. run `make ci-go-flake`

It'll run until it fails. Keep it running for some time to convince yourself that it's less likely to fail than before. (On master, it doesn't take long on my machine.)

-------
Fixes #3272

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -1502,6 +1502,7 @@ func (s *Server) v1DataPost(w http.ResponseWriter, r *http.Request) {
 			result.Explanation, err = types.NewTraceV1(*buf, pretty)
 			if err != nil {
 				writer.ErrorAuto(w, err)
+				return
```
Comment (srenatus) [inline server/server.go:1505]: This is unrelated to the change, something I've stumbled upon. I don't believe there should be any `writer.ErrorAuto` calls that aren't followed by a `return`.

**Review by tsandall — DISMISSED** [review]
Mostly just questions. Took me a while to understand how callOrCancel works but that's my own fault for looking at this at the end of the day 😅 

I'm looking forward to getting this merged and hopefully saying goodbye to flakes!

File: internal/wasm/sdk/internal/wasm/vm.go [hunk 2]
```diff
@@ -24,6 +24,19 @@ import (
...
+// to them. We've seen that under CI conditions the benchmarks
+// would be flaky -- they run this code many many times and have
+// low resources. With n retries, it's still possible that each
+// try gets some interrupt that didn't belong to it -- but the
+// higher n is, the less likely it becomes.
+// With n=1, I've seen the failures locally very rarely. So,
+// n=2 should be fine. If it turns out not to be, we can adjust
+// either this constant or our strategy for dealing with
+// interrupts.
+const maxInterruptRetries = 2
```
[the remark this answers is not in the document]
Comment (srenatus) [reply internal/wasm/sdk/internal/wasm/vm.go:38]: Oh. Of course :D

File: internal/wasm/sdk/internal/wasm/vm.go [hunk 3]
```diff
@@ -605,45 +585,82 @@ func (i *VM) cloneDataSegment() (int32, []byte) {
...
-		// trap of a previous evaluation.
-		select {
-		case <-ctx.Done(): // chan closed
-			return errors.ErrCancelled
-		default: // don't block
+
+	select {
+	case <-once(ctx.Done()):
+		// interrupt, wait for trap
+		cancelled = true
```
[the remark this answers is not in the document]
Comment (srenatus) [reply internal/wasm/sdk/internal/wasm/vm.go:647]: Nope another leftover I've been meaning to clean up. Only matters if called in a loop, that doesn't happen here.

File: internal/wasm/sdk/internal/wasm/vm.go [hunk 4]
```diff
@@ -605,45 +585,82 @@ func (i *VM) cloneDataSegment() (int32, []byte) {
...
+			return 0, err
 		}
-		return nil
+		return res, nil
 	}
-	return err
+	// clear trap: reaching this, we have definitely been interrupted
+	err := <-errc
+	for err != errors.ErrCancelled && !trap(err) {
+		_, err = vm.instance.GetExport("opa_heap_ptr_get").Func().Call() // cheap call
```
[the remark this answers is not in the document]
Comment (srenatus) [reply internal/wasm/sdk/internal/wasm/vm.go:658]: Yup it's in the for loop's condition

File: internal/wasm/sdk/internal/wasm/vm.go [hunk 5]
```diff
@@ -322,15 +346,13 @@ func (i *VM) Eval(ctx context.Context, entrypoint int32, input *interface{}, met
 
 	// Skip free'ing input and result JSON as the heap will be reset next round anyway.
 
-	return data[0:n], err
+	return data[0:n], err // err?!
```
[the remark this answers is not in the document]
Comment (srenatus) [reply internal/wasm/sdk/internal/wasm/vm.go:316]: Ah sorry, I'll update. I don't see how this would ever be non nil here.

File: internal/wasm/sdk/internal/wasm/vm.go [hunk 6]
```diff
@@ -605,45 +585,82 @@ func (i *VM) cloneDataSegment() (int32, []byte) {
...
+						errc <- nil
+						return
+					}
+					errc <- e.err
+				default:
+					panic(e)
+				}
+			}
+		}()
+		res, err = vm.instance.GetExport(name).Func().Call(sl...)
```
[the remark this answers is not in the document]
Comment (srenatus) [reply internal/wasm/sdk/internal/wasm/vm.go:619]: Yeah: https://github.com/bytecodealliance/wasmtime-go/blob/4d860e263dc6ff070388d365c3e6b59e981df4b8/func.go#L500-L514
